### PR TITLE
WIP: orchestrator: distribuited: NatsOrchestrator 

### PR DESCRIPTION
**Pull request by aghinsa**
Milestone 1 of #790

### CODE REVIEW

File: dffml/distribuited/orchestrator.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+
+
+# Subjects
+RegisterWorkerNode = "RegisterWorkerNode"
+ConnectToOrchestratorNode = "ConnectToOrchestratorNode"
+OrchestratorNodeReady = "OrchestratorNodeReady"
+
+####
+
+NBYTES_UID = 6
```
Comment (pdxjohnny) [inline dffml/distribuited/orchestrator.py:41]: Let's take this out of the http service and put it in util somewhere, maybe `dffml/util/default.py` then use it in this file

https://github.com/intel/dffml/blob/ad2286448b18ccd2b844e01ce02b8c1712d02e8d/service/http/dffml_service_http/routes.py#L41-L44

File: tests/test_distribuited/test_nats.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+class TestNatsOrchestrator(AsyncTestCase):
+    async def setUp(self):
+        with socketserver.TCPServer(("localhost", 0), None) as s:
+            free_port = s.server_address[1]
+
+        self.server_addr = f"0.0.0.0:{free_port}"
+
+        self.nats_proc = subprocess.Popen(
+            ["nats-server", "-p", str(free_port),],
```
Comment (pdxjohnny) [inline tests/test_distribuited/test_nats.py:50]: Let's make sure we have NATs choose the port

https://github.com/nats-io/nats-server/issues/1097#issuecomment-637232781

```suggestion
            ["nats-server", "-p", "-1"],
```

File: dffml/distribuited/orchestrator.py [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+            f"{RegisterWorkerNode}.{self.onid}", msg_data
+        )
+        # Response contains operation names mapped to token number
+        response = json.loads(response.data.decode())
+        self.operation_token.update(response)
+        self.logger.debug(f"Tokens received: {response}")
+
+        # Look for implementations provided in sub node:
+        for operation_name in response:
+            if operation_name in self.parent.config.implementations:
```
Comment (pdxjohnny) [inline dffml/distribuited/orchestrator.py:382]: The else case here is not handled, what is the else case?

File: tests/test_distribuited/test_nats.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                    operations={
+                        "check_url": check_if_valid_git_repository_URL,
+                        "dbq": db_query_create_table.op,
+                    },
+                    configs={"dbq": DatabaseQueryConfig(database=sdb)},
+                ),
+            )
+        )
+
+        sn1 = await subnode_1.__aenter__()
```
Comment (pdxjohnny) [inline tests/test_distribuited/test_nats.py:93]: Why no use of `with` here?

File: dffml/distribuited/orchestrator.py (hunk 3 above)
Comment (aghinsa) [reply dffml/distribuited/orchestrator.py:382]: If no implementations are provided, the operation is assumed to be entrypoint loadable. opimpNetworkctx raises an exception if an implementation is not found in both ways.

File: tests/test_distribuited/test_nats.py (hunk 4 above)
Comment (aghinsa) [reply tests/test_distribuited/test_nats.py:93]: My bad, updated to use with

File: dffml/distribuited/notes.md [hunk 5]
```diff
@@ -0,0 +1,24 @@
+## Notes
```
Comment (pdxjohnny) [inline dffml/distribuited/notes.md:1]: Let's move this into `docs/`

File: dffml/distribuited/orchestrator.py [hunk 6]
```diff
@@ -0,0 +1,431 @@
...
+
+class NatsNode(BaseDataFlowObject):
+    async def error_handler(self, msg):
+        tb = traceback.format_exc()
+        print()
+        print(tb)
+        print(f"ERROR in node {self.uid}")
+        print()
+
+    async def __aenter__(self, enter={}):
```
Comment (pdxjohnny) [inline dffml/distribuited/orchestrator.py:125]: Let's remove arguments from all `__aenter__`'s

```suggestion
    async def __aenter__(self):
```

File: tests/test_distribuited/test_nats.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                self.server_addr = line[-1]
+            if "Server is ready" in line:
+                ready = True
+
+    async def tearDown(self):
+        self.nats_proc.terminate()
+
+    async def test_run(self):
+        server = self.server_addr
+        subnode_1 = NatsWorkerNode(
```
Comment (pdxjohnny) [inline tests/test_distribuited/test_nats.py:66]: Let's instantiate 3 workers (configurable to N workers, maybe use a loop or something) and make sure that it still work.

Let's give each a different set of operations and make some of them overlap and some unique and makes sure that circular queues are setup appropriately.

File: dffml/distributed/orchestrator.py [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+        self.required_operations = [
+            operation.name
+            for operation in self.parent.config.dataflow.operations.values()
+        ]
+
+        # maps operations to registered nodes token queue
+        self.nodes_for_operation = (
+            {}
+        )
+
```
Comment (pdxjohnny) [inline dffml/distributed/orchestrator.py:212]:
```suggestion
        # Workers will register themselves with this orchestrator
        # by publishing to the following channel
```

File: dffml/distributed/orchestrator.py [hunk 9]
```diff
@@ -0,0 +1,430 @@
...
+        sid_node_registration = await self.nc.subscribe(
+            f"{RegisterWorkerNode}.{self.uid}",
+            queue="WorkerNodeRegistrationQueue",
+            cb=self.register_worker_node,
+        )
+        self.logger.debug("Waiting for all operations to be found")
+        await self.all_ops_available.wait()
+        self.logger.debug("All required operations found")
+        # Stop listening to new node registration
+        await self.nc.unsubscribe(
```
Comment (pdxjohnny) [inline dffml/distributed/orchestrator.py:222]: TODO: remove this line

File: dffml/distributed/orchestrator.py [hunk 10]
```diff
@@ -0,0 +1,430 @@
...
+        # in dataflow and assign an instance to a worker node
+        # making sure that each worker node gets at least
+        # one instance
+        self.dataflow = self.parent.config.dataflow
+        self.node_token_managing_instance = {
+            instance_name: (
+                operation.name,
+                self.nodes_for_operation[operation.name].get(),
+            )
+            for instance_name, operation in self.dataflow.operations.items()
```
Comment (pdxjohnny) [inline dffml/distributed/orchestrator.py:236]: Move this before the `.wait()`

File: dffml/distributed/todos.md [hunk 11]
```diff
@@ -0,0 +1,12 @@
+## ToDo
+Onode waits till it has all the operations in the dataflow
+    - If there are multiple nodes having the same operation, but
+    for some reason their ack does not reach Onode before it has
+    obtained acks for all the operations, those worker nodes won't
+    be associated with the context.
```
Comment (pdxjohnny) [inline dffml/distributed/todos.md:6]: Please make sure we have a `README.md` in this directory that combines your nodes and todos and is up to date with the state of the distributed implementations, it's limitations, and reasonings for those limitations

**Pull request closed without merging**